packages/core/src/modules/action-menu/services/ActionMenuService.ts

```diff
@@ -0,0 +1,282 @@
+import type { InboundMessageContext } from '../../../agent/models/InboundMessageContext'
+import type { Logger } from '../../../logger'
+import type { ActionMenuStateChangedEvent } from '../ActionMenuEvents'
+import type { FindActiveMenuOptions } from '../ActionMenuModuleOptions'
+import type { CreateMenuOptions, CreatePerformOptions, CreateRequestOptions } from './ActionMenuServiceOptions'
+
+import { AgentConfig } from '../../../agent/AgentConfig'
+import { EventEmitter } from '../../../agent/EventEmitter'
+import { AriesFrameworkError } from '../../../error'
+import { injectable } from '../../../plugins'
+import { JsonTransformer } from '../../../utils'
+import { ActionMenuEventTypes } from '../ActionMenuEvents'
+import { ActionMenuRole } from '../ActionMenuRole'
+import { ActionMenuState } from '../ActionMenuState'
+import { PerformMessage, MenuMessage, MenuRequestMessage } from '../messages'
+import { ActionMenuSelection, ActionMenu } from '../models'
+import { ActionMenuRepository, ActionMenuRecord } from '../repository'
+
+@injectable()
+export class ActionMenuService {
+  private actionMenuRepository: ActionMenuRepository
+  private eventEmitter: EventEmitter
+  private logger: Logger
+
+  public constructor(actionMenuRepository: ActionMenuRepository, agentConfig: AgentConfig, eventEmitter: EventEmitter) {
+    this.actionMenuRepository = actionMenuRepository
+    this.eventEmitter = eventEmitter
+    this.logger = agentConfig.logger
+  }
+
+  public async createRequest(options: CreateRequestOptions) {
+    // Assert
+    options.connection.assertReady()
+
+    // Create message
+    const menuRequestMessage = new MenuRequestMessage({})
+
+    // Create record if not existant for connection/role
+    let actionMenuRecord = await this.findActive({
+      connectionId: options.connection.id,
+      role: ActionMenuRole.Requester,
+    })
+
+    if (actionMenuRecord) {
+      const previousState = actionMenuRecord.state
+      actionMenuRecord.state = ActionMenuState.AwaitingRootMenu
+      actionMenuRecord.threadId = menuRequestMessage.id
+
+      await this.actionMenuRepository.update(actionMenuRecord)
+      this.emitStateChangedEvent(actionMenuRecord, previousState)
```

Comment:
What if the action menu is currently in progress? Can I just start a new one? 

I see you mentioned: 
> As this protocol currently defines a single active menu per connection, there will be at most 2 records per connection (as theoretically is possible to serve as requester and responder at the same time).

Is this a limitation imposed by the RFC?

Reply:
From the RFC I understood that this situation would discard current flow and request back the root menu:
> At any time a requester can reset the protocol by requesting the root menu from a responder.

The RFC states also that:

> both Agents may support both requester and responder roles simultaneously. Such cases would result in two instances of the action-menu protocol operating in parrallel.

So that's why we are supporting one instance per role and connection. At first I thought to create an ActionMenuRecord per interaction (for instance, every time the requester agent selects an option, it creates a new record, which would be set to Done once a new menu is requested and then another record is created). But it could become a bit too complicated considering the simplicity of current specification.